ase/sw/ase_host_memory.c

```diff
@@ -0,0 +1,430 @@
+	// for kernel addresses.  This simulation is user space.
+	if ((uint64_t) va >> 48) {
+		ASE_ERR("Unexpected virtual address (0x" PRIx64 "), high bits set!", va);
+		raise(SIGABRT);
+	}
+
+	int pt_level = ase_pt_length_to_level(length);
+	uint64_t mask = 0xffffffffffffUL & (~0UL << (12 + 9 * pt_level));
+	return (uint64_t) va ^ mask;
+}
+
+/*
+ * Translate from simulated physical address space. Optionally hold the lock
+ * after successful translation so that the buffer remains pinned.
+ */
+void *ase_host_memory_pa_to_va(uint64_t pa, bool lock)
+{
+	// The inverse of ase_host_memory_va_to_pa.
+	if (pa >> 48) {
+		ASE_ERR("Unexpected physical address (0x" PRIx64 "), high bits set!", pa);
+		raise(SIGABRT);
+	}
+
+	// Is the page pinned?
+	int pt_level;
+	pthread_mutex_lock(&ase_pt_lock);
+	if (!ase_pt_check_addr(pa, &pt_level)) {
+		pthread_mutex_unlock(&ase_pt_lock);
+		return NULL;
+	}
+
+	if (!lock) {
+		pthread_mutex_unlock(&ase_pt_lock);
+	}
+
+	uint64_t mask = 0xffffffffffffUL & (~0UL << (12 + 9 * pt_level));
+	return (void *) (pa ^ mask);
+}
+
+
+void ase_host_memory_unlock(void)
+{
+	pthread_mutex_unlock(&ase_pt_lock);
+}
+
+
+void ase_host_memory_terminate(void)
+{
+	pthread_mutex_lock(&ase_pt_lock);
+
+	ase_pt_delete_tree(ase_pt_root, 3);
+	ase_pt_root = NULL;
+
+	pthread_mutex_unlock(&ase_pt_lock);
+}
+
+
+// ========================================================================
+//
+//	Maintain a simulated page table in order to track pinned pages.
+//
+// ========================================================================
+
+/*
+ * The page table here has a single job: indicate whether a given physical
+ * address (iova) is pinned. Since the simulator uses a simple XOR function
+ * for mapping virtual to physical addresses, the table is not needed for
+ * actual translation.
+ *
+ * The table is physically indexed but the internal pointers are simple
+ * virtual addresses. At the lowest level (the 4KB pages), the table uses
+ * a 512 entry bit vector to indicate whether a page is pinned. This saves
+ * space and is, once again, possible since the table isn't used for
+ * actual translation.
+ *
+ * A pointer value of -1 at higher levels in the table indicates the
+ * presence of a pinned huge page.
+ */
+
+/*
+ * Page size to level in the table. Level 3 is the root.
+ */
+static int ase_pt_length_to_level(uint64_t length)
+{
+	int pt_level;
+
+	if (length > 2 * MB)
+		pt_level = (length == GB) ? 2 : -1;
+	else if (length > 4 * KB)
+		pt_level = (length <= 2 * MB) ? 1 : -1;
+	else
+		pt_level = 0;
+
+	return pt_level;
+}
+
+/*
+ * Index of a 512 entry set of page pointers in the table, given a level.
+ */
+static inline int ase_pt_idx(uint64_t iova, int pt_level)
+{
+	// The low 12 bits are the 4KB page offset. Groups of 9 bits above that
+	// correspond to increasing levels in the page table hierarchy.
+	assert(pt_level <= 3);
+	return 0x1ff & (iova >> (12 + pt_level * 9));
```

Comment:
Neat. but, requires some effort to understand. consider replacing 12 and 9 with LOG2_PAGE_SIZE, LOG2_GROUP_SIZE or some other meaningful macro